### PR TITLE
Refine automatic mixed precision support

### PR DESCRIPTION
**Pull request by vinhngx**
In continuation of  https://github.com/tensorflow/tensor2tensor/pull/1637

In this PR, we re-organize automatic mixed precision training support to provide a cleaner implementation and an easier interface. 

In particular, [GPU automatic mixed precision training](https://medium.com/tensorflow/automatic-mixed-precision-in-tensorflow-for-faster-ai-training-on-nvidia-gpus-6033234b2540) can now be enabled via setting the flag `--gpu_automatic_mixed_precision` for **all tensor2tensor models**, for example: 

**Transformer**
```
PROBLEM=translate_ende_wmt32k
MODEL=transformer
HPARAMS=transformer_big
DATA_DIR=/data/translate_ende_wmt32k
TRAIN_DIR=/tmp/$MODEL-$HPARAMS

t2t-trainer \
  --data_dir=$DATA_DIR \
  --problem=$PROBLEM \
  --model=$MODEL \
  --hparams_set=$HPARAMS \
  --output_dir=$TRAIN_DIR \
  --train_steps=100000 \
  --eval_steps=1000 \
  --gpu_automatic_mixed_precision=True
```

**Resnet:**
```
PROBLEM=image_imagenet224
MODEL=resnet
HPARAMS=resnet_50
DATA_DIR=/data/ImageNet
TRAIN_DIR=/tmp/$HPARAMS

t2t-trainer \
  --data_dir=$DATA_DIR \
  --problem=$PROBLEM \
  --model=$MODEL \
  --hparams_set=$HPARAMS \
  --output_dir=$TRAIN_DIR \
  --hparams='batch_size=256' \
  --worker_gpu=8 \
  --gpu_automatic_mixed_precision=True
```

This is opposed to the previous approaches of setting the OS flag `TF_ENABLE_AUTO_MIXED_PRECISION` which is a non-programatic approach, or passing the flag `gpu_auto_mixed_precision` directly to the optimizer (which will require modification of individual models to make call to optimizer with mixed precision training option).

### CODE REVIEW

File: tensor2tensor/utils/optimize.py [hunk 1]
```diff
@@ -71,8 +69,7 @@ def optimize(loss,
   opt = ConditionalOptimizer(hparams.optimizer, learning_rate, hparams, use_tpu)
   if use_tpu:
     opt = tf.contrib.tpu.CrossShardOptimizer(opt)
-  if gpu_auto_mixed_precision or os.environ.get(
-      "TF_ENABLE_AUTO_MIXED_PRECISION", "0") == "1":
+  if tf.flags.FLAGS.gpu_automatic_mixed_precision:
```
[the remark this answers is not in the document]
Comment (vinhngx) [reply tensor2tensor/utils/optimize.py]: hi @afrozenator  thanks for the feedback. I looked into `hparams` and to be honest got lost a bit in that maze :) Using a flag seems to result in much simpler and cleaner implementation. But if you insist using `hparams` I'd appreciate some pointers here

[the remark this answers is not in the document]
Comment (vinhngx) [reply tensor2tensor/utils/optimize.py]: @afrozenator  have a look at https://github.com/tensorflow/tensor2tensor/pull/1681 for an alternative as per your suggestion

**Comment by afrozenator** [comment]
Thanks @vinhngx - #1681 certainly looks better than this, but we might want to keep the t2t_trainer.py flag from here and set the hparam property based on that flag -- does this look good to you?

**Comment by vinhngx** [comment]
yeah sounds good. Let me have a look into that.

**Comment by vinhngx** [comment]
close in favor of #1681

**Pull request closed without merging**